hs_access_control/models.py

```diff
@@ -77,27 +84,141 @@ class PrivilegeCodes(object):
     CHANGE = 2
     VIEW = 3
     NONE = 4
-    PRIVILEGE_CHOICES = (
+    CHOICES = (
         (OWNER, 'Owner'),
         (CHANGE, 'Change'),
         (VIEW, 'View')
         # (NONE, 'None') : disallow "no privilege" lines
     )
     # Names of privileges for printing
-    PRIVILEGE_NAMES = ('Unspecified', 'Owner', 'Change', 'View', 'None')
+    NAMES = ('Unspecified', 'Owner', 'Change', 'View', 'None')
+
+
+class PrivilegeBase(models.Model):
+    """
+    Shared methods for Privilege handling
+
+    These polymorphic routines act on any type of foreign keys, so that
+    they are independent of the types of the base class members.
+    They implement the major interactions with privileges, including
+    share, unshare, undo_share, and ensure that provenance and privilege
+    are always in sync with one another.
+    """
+    class Meta:
+        abstract = True
+
+    @classmethod
+    def get_privilege(cls, **kwargs):
+        """
+        Get an effective privilege for a pair of keys
+
+        This works for any pair of attributes that together, form a key.
+        Examples include:
+
+            * UserResourcePrivilege.get_privilege(user={X}, resource={Y})
+            * GroupResourcePrivilege.get_privilege(group={X}, resource={Y})
+            * UserGroupPrivilege.get_privilege(user={X}, group={Y})
+
+        **This is a system routine** and not recommended for use in application code.
+        """
+        try:
+            return cls.objects.get(**kwargs).privilege
+        except cls.DoesNotExist:
+            return PrivilegeCodes.NONE
+
+    @classmethod
+    def update(cls, **kwargs):
+        """
+        Update an effective privilege record without maintaining provenance.
+
+        This works for any pair of attributes that together, form a key, along with shared
+        parameters `grantor` and `privilege`.  Examples include:
+
+            * UserResourcePrivilege.update(user={X}, resource={Y}, privilege={Z}, grantor={W})
+            * GroupResourcePrivilege.update(group={X}, resource={Y}, privilege={Z}, grantor={W})
+            * UserGroupPrivilege.update(user={X}, group={Y}, privilege={Z}, grantor={W})
+
+        **This is a system routine** and not recommended for use in application code.
+        There are no access control rules applied; this routine is unconditional.
+        Only use this routine if you wish to completely bypass access control.
+        Note also that using this routine directly breaks provenance and disables undo.
+        """
+        grantor = kwargs['grantor']
+        privilege = kwargs['privilege']
```

Comment:
It seems ```privilege``` can be ```None``` which means this line can throw ```KeyError``` exception. You may want to do:

```privilege = kwargs.get('privilege', None)```